### PR TITLE
route_generator

### PR DESCRIPTION
**Pull request by Inno97**
Create the various RouteNode classes and automate data population for them.
Create a command RouteGenerateCommand, that creates a Route from 2 input locations.

Features:

	 - Create auto-updating function in BusStop and TrainStation
		 - Implement fetchData in BusStop and TrainStation, taking data from local storage
	 
	 - Add TrainStation class
	 
	 - Add CustomNode class 
		 - For RouteNodes that are not BusStops or TrainStations
	 
	 - Modify PathFinder to cast to Bus Stop and Train Station automatically
	 
	 - Make getInbetweenNodes method
		 - To be called when nodes are of the same type
		 - Returns the intermediate RouteNodes between these 2 points
	 
	 - Add RouteGenerateCommand
		 - Generates a Route based on 2 locations and constraint
		 - Includes all RouteNodes in the path
		 - Adds to RouteList after creation
		 - CustomNodes are added inbetween as instructions for the routes
	 
	 - Add relevant Conversations for Routes
		 - RouteAddConversation
		 - RouteDeleteConversation
		 - RouteEditConversation
		 - RouteGenerateConversation
		 - RouteListConversation
		 - RouteNodeAddConversation
		 - RouteNodeDeleteConversation
		 - RouteNodeEditConversation
		 - RouteNodeListConversation
		 
	 - Standardise the usage of getIndex-related methods in parsing
		 - Always decrement in these methods and not in commands

	 - Refactor methods in RouteNode and related classes
		 - Removed unnecessary method declarations in package duke.model.locations.
		 
		 
	 
	 - Add tests for the following classesTest
		 - RouteAddConversationTest
		 - RouteDeleteConversationTest
		 - RouteEditConversationTest
		 - RouteGenerateConversationTest
		 - RouteListConversationTest
		 - RouteNodeAddConversationTest
		 - RouteNodeDeleteConversationTest
		 - RouteNodeEditConversationTest
		 - RouteNodeListConversationTest
	 
Fixes:

	 - Fix RouteNodeAddCommand to catch exception when queries Route does not exist.
		 - Example: routeNodeAdd 0 0 at 17009 by bus exceptions are not caught.

### CODE REVIEW

File: routes.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+route | test |
```
Comment (Jefferson111) [inline routes.txt]: Delete this

Comment (Inno97) [reply routes.txt]: :d
Yeah i keep forgetting about this.

File: src/main/java/duke/commons/enumerations/Constraint.java [hunk 2]
```diff
@@ -7,5 +7,6 @@
     MRT,
     BUS,
     MIXED,
-    CAR;
+    CAR,
+    CUSTOM;
```
Comment (Jefferson111) [inline src/main/java/duke/commons/enumerations/Constraint.java]: Not sure if we need add ```;``` in enums

Comment (Inno97) [reply src/main/java/duke/commons/enumerations/Constraint.java]: Checked, we do not need ; in enums.

File: src/main/java/duke/commons/exceptions/ObjectCreationFailedException.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package duke.commons.exceptions;
+
+import duke.commons.Messages;
+
+/**
+ * Exception thrown when index query is out of bounds.
+ */
```
Comment (Jefferson111) [inline src/main/java/duke/commons/exceptions/ObjectCreationFailedException.java]: Description need update?

File: src/main/java/duke/commons/exceptions/QueryFailedException.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package duke.commons.exceptions;
+
+import duke.commons.Messages;
+
+/**
+ * Exception thrown when index query is out of bounds.
+ */
```
Comment (Jefferson111) [inline src/main/java/duke/commons/exceptions/QueryFailedException.java]: The description here too?

File: src/main/java/duke/logic/api/ApiParser.java [hunk 5]
```diff
@@ -10,6 +10,7 @@
 import duke.logic.api.requests.DataMallHttpRequest;
 import duke.logic.api.requests.StaticMapUrlRequest;
 import duke.model.locations.BusStop;
+import duke.model.locations.RouteNode;
 import duke.model.locations.Venue;
 import duke.model.transports.BusService;
```
Comment (Jefferson111) [inline src/main/java/duke/logic/api/ApiParser.java]: Separate different imports

File: src/main/java/duke/logic/conversations/Conversation.java [hunk 6]
```diff
@@ -44,33 +46,85 @@ public String getPrompt() {
...
     /**
-     * Checks and sets input if is dateTime.
-     * @param input The input
-     * @return String The input if is dateTime
+     * Checks if input is a field of a Route.
+     *
+     * @param input The input.
+     * @return If the input is a field of a Route.
+     */
+    protected boolean isRouteField(String input) {
```
Comment (Jefferson111) [inline src/main/java/duke/logic/conversations/Conversation.java]: This does not belong to conversation

File: src/main/java/duke/logic/conversations/Conversation.java [hunk 7]
```diff
@@ -44,33 +46,85 @@ public String getPrompt() {
...
     /**
-     * Checks and sets input if is int.
-     * @param input The input
-     * @return String The input if is int
+     * Checks if input is int.
+     *
+     * @param input The input.
+     * @return If the input is int.
      */
     protected Boolean isIntInput(String input) {
```
Comment (Jefferson111) [inline src/main/java/duke/logic/conversations/Conversation.java]: We should extract all of these check conditions into one class each

File: src/main/java/duke/logic/parsers/ParserUtil.java [hunk 8]
```diff
@@ -1,19 +1,28 @@
 package duke.logic.parsers;
```
Comment (Jefferson111) [inline src/main/java/duke/logic/parsers/ParserUtil.java]: Need to breakdown this class

File: src/main/java/duke/model/Model.java [hunk 9]
```diff
@@ -13,6 +15,7 @@
 import duke.model.transports.BusService;
 import duke.model.locations.BusStop;
 import duke.model.locations.Venue;
+import duke.model.transports.Route;
 
 import java.util.List;
```
Comment (Jefferson111) [inline src/main/java/duke/model/Model.java]: No idea why Github is showing me the old model here, but please check if ur side is the updated one

File: src/main/java/duke/logic/conversations/Conversation.java (hunk 7 above)
Comment (Inno97) [reply src/main/java/duke/logic/conversations/Conversation.java]: Extracted into ConversationUtil.java.

File: src/main/java/duke/logic/parsers/ParserUtil.java (hunk 8 above)
Comment (Inno97) [reply src/main/java/duke/logic/parsers/ParserUtil.java]: Will be done in 1.4.

File: src/main/java/duke/model/Model.java (hunk 9 above)
Comment (Inno97) [reply src/main/java/duke/model/Model.java]: Cross-checked with current build. This PR contains an updated one.

File: src/main/java/duke/logic/conversations/Conversation.java (hunk 6 above)
Comment (Inno97) [reply src/main/java/duke/logic/conversations/Conversation.java]: Extracted into ConversationUtil.java.

**Comment by Inno97** [comment]
There's still the "Use explicit scoping instead of the default package private level" in Codacy.
For the life of me, I don't know how to fix it.

**Pull request merged**